src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -106,14 +106,23 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
                     }
                     catch (Exception e)
                     {
-                         // We did not finish sending body so notify server.
+                         // Try to notify server if we did not finish sending request body.
                          _ = _connection.SendRstStreamAsync(_streamId, Http2ProtocolErrorCode.Cancel);
 
                         // if we decided abandon sending request and we get ObjectDisposed as result of it, just eat exception.
-                        if (_shouldSendRequestBody || (!(e is ObjectDisposedException) && !(e.InnerException is ObjectDisposedException)))
+                        if (!_shouldSendRequestBody && (e is ObjectDisposedException || e.InnerException is ObjectDisposedException))
                         {
-                            throw;
+                            return;
+                        }
+
+                        if (_abortException == null)
+                        {
+                            // If we still processing response after receiving response headers, this will give us chance to propagate exception up.
+                            // Since we failed while Copying stream, wrapp it as IOException if needed.
+                            _abortException = e;
```

Comment:
It looks like we could end up trying to race to set _abortException.  Presumably if there are multiple exceptions, we don't care which one wins?  Or are there situations where we should be prioritizing one over another?

Reply:
I don't know about any case where it would matter. Maybe @geoffkizer can comment on it but my assumption was that first exception would be closer to root cause and second would maybe be result of it. I guess we could set them atomically and nest but I'm not sure if that would be useful.

Reply:
I'm honestly not sure how _abortException is used. Can someone explain what this is for?

Reply:
That is mechanism how to propagate errors from background tasks  - like ProcessIncomingFramesAsync(). When exception happens we save it and than later on we check if something happened. In this case when response headers are already given back to caller, we use that to propagate errors from sending request body (of still pending) to response. This is streategy we discussed while back for the duplex send/receive logic.

Reply:
Ok. So this will be something like an Http2ProtocolException or an IOException, etc? And then it becomes the inner exception for HttpRequestException somewhere?

I guess my main question here is, why are we setting this here? Seems like it should be set at a lower level, in one of the connection read or write methods.

Reply:
in general, you are right.  For exceptions caused by IO or protocol, this should be already set. This was raised in context of the concurrent  body send  while receiving response headers and response body. If the body send fails for reasons like OperationCanceled or ObjectDisposed or some other reasons, this ensures we will be able to propagate that  error while processing response body. This is generally fall-back for cases when we already process response headers and return them to caller together with response stream.